src/Backups/BackupsWorker.cpp

```diff
@@ -70,365 +120,407 @@ UUID BackupsWorker::start(const ASTPtr & backup_or_restore_query, ContextMutable
+            restoreTablesData(std::move(data_restore_tasks), restores_thread_pool);
 
-                /// We have restored everything, we need to tell other hosts (they could be waiting for it).
-                restore_coordination->setStatus(restore_settings.host_id, kCompletedCoordinationStatus, "");
-            }
+            /// We have restored everything, we need to tell other hosts (they could be waiting for it).
+            restore_coordination->setStage(restore_settings.host_id, kCompletedStage, "");
+        }
 
-            setStatus(restore_uuid, BackupStatus::RESTORED);
+        LOG_INFO(log, "Restored from {} {} successfully", (restore_settings.internal ? "internal backup" : "backup"), backup_info.toString());
+        setStatus(restore_uuid, restore_settings.internal, BackupStatus::RESTORED);
+    }
+    catch (...)
+    {
+        /// Something bad happened, the backup has not built.
+        if (called_async)
+        {
+            tryLogCurrentException(log, fmt::format("Failed to restore from {} {}", (restore_settings.internal ? "internal backup" : "backup"), backup_info.toString()));
+            setStatus(restore_uuid, restore_settings.internal, BackupStatus::FAILED_TO_RESTORE);
+            sendCurrentExceptionToCoordination(restore_coordination, restore_settings.host_id);
         }
-        catch (...)
+        else
         {
-            /// Something bad happened, the backup has not built.
-            setStatus(restore_uuid, BackupStatus::FAILED_TO_RESTORE);
-            sendErrorToCoordination(restore_coordination, restore_settings.host_id);
-            if (!async)
-                throw;
+            /// setStatus() and sendCurrentExceptionToCoordination() will be called by startRestoring().
+            throw;
         }
-    };
-
-    if (restore_settings.async)
-        backups_thread_pool.scheduleOrThrowOnError([job]() mutable { job(true); });
-    else
-        job(false);
-
-    return restore_uuid;
+    }
 }
 
 
-void BackupsWorker::addInfo(const UUID & uuid, const String & backup_name, BackupStatus status, bool internal)
+void BackupsWorker::addInfo(const UUID & uuid, bool internal, const String & backup_name, BackupStatus status)
 {
     Info info;
     info.uuid = uuid;
     info.backup_name = backup_name;
     info.status = status;
     info.status_changed_time = time(nullptr);
     info.internal = internal;
+
     std::lock_guard lock{infos_mutex};
-    infos[uuid] = std::move(info);
+    infos[{uuid, internal}] = std::move(info);
+
+    num_active_backups += getNumActiveBackupsChange(status);
+    num_active_restores += getNumActiveRestoresChange(status);
 }
 
-void BackupsWorker::setStatus(const UUID & uuid, BackupStatus status)
+
+void BackupsWorker::setStatus(const UUID & uuid, bool internal, BackupStatus status)
 {
     std::lock_guard lock{infos_mutex};
-    auto & info = infos.at(uuid);
+    auto it = infos.find({uuid, internal});
+    if (it == infos.end())
+        return;
```

Comment:
When is it possible? Why not to throw `LOGICAL_ERROR`?

Reply:
Replaced with LOGICAL_ERROR